src/utils/recompiler.ts

```diff
@@ -0,0 +1,132 @@
+        hash.update(data);
+        newNameHashPair[cairoContract] = hash.digest("hex").toString();
+    }
+
+    return newNameHashPair;
+};
+
+// Checks artifacts availability
+const checkArtifacts = async (
+    hre: HardhatRuntimeEnvironment,
+    newNameHashPair: NameHashPair,
+    changed: Set<string>
+): Promise<void> => {
+    const { starknetSources: defaultSourcesPath } = hre.config.paths;
+    const sourceRegex = new RegExp("^" + defaultSourcesPath + "/");
+
+    const artifactsDir = getArtifactPath(defaultSourcesPath, hre);
+    const artifactsPathRegex = new RegExp("^" + artifactsDir + "/");
+    // traverse on artifacts directory
+    let artifactsDirList = await traverseDirectories(artifactsDir, "*.cairo");
+    artifactsDirList = artifactsDirList.map(file => file.replace(artifactsPathRegex, ""));
+
+    for (const name in newNameHashPair) {
+        const filePath = name.replace(sourceRegex, "");
+        if (!artifactsDirList.includes(filePath)) {
+            changed.add(name);
+        }
+    }
+};
+
+// Compile changed contracts
+const compileChangedContracts = async (
+    args: TaskArguments,
+    hre: HardhatRuntimeEnvironment,
+    newNameHashPair: NameHashPair,
+    oldNameHashPair: NameHashPair,
+    changed: Set<string>
+): Promise<void> => {
+    for (const contractName in newNameHashPair) {
+        // Add new contracts that are not in cache before
+        if (!oldNameHashPair[contractName]) {
+            changed.add(contractName);
+        }
+
+        // Add contracts that contiain a change in content
+        if (newNameHashPair[contractName] !== oldNameHashPair[contractName]) {
+            changed.add(contractName);
+        }
+    }
+
+    if (changed.size > 0) {
+        // Compiles contracts
+        console.log("Compiling contracts...");
+        args.paths = [...changed];
+        await starknetCompileAction(args, hre);
+    }
+};
+
+export const handleCache = async (args: TaskArguments, hre: HardhatRuntimeEnvironment): Promise<boolean> => {
+    // If recompile is not enabled, do nothing
+    if (!hre.userConfig?.starknet?.recompile) return true;
+
+    // Get cache directory, default source directory and artifacts directory from hre.config
+    const { starknetSources: defaultSourcesPath, cache: cacheDirName } = hre.config.paths;
+
+    // Set to save contracts with changed content & unavailable artifacts
+    const changed: Set<string> = new Set();
+
+    try {
+        const oldNameHashPair = await upsertFile(cacheDirName);
+        const newNameHashPair = await getContractHash(defaultSourcesPath);
+        await checkArtifacts(hre, newNameHashPair, changed);
+        await compileChangedContracts(args, hre, newNameHashPair, oldNameHashPair, changed);
+
+        // Write to file new NameHashPair of contracts
+        const cacheFile = path.join(cacheDirName, CACHE_FILE_NAME);
+        await fsPromises.writeFile(cacheFile, JSON.stringify(newNameHashPair, null, " "));
```

Comment:
As discussed, saving of the cache file should be done in starknetCompileAction (i.e. the starknet-compile command should also take care of caching - by just generating the file)

Reply:
https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/106/commits/cf0f0bbd5cec174e5cbd0fafe517d2589d5a8bd2